File: 3.math-stats/all-combinations/combine.py
# leetcode: https://leetcode.com/problems/combinations/
def combine(n, k):
    res = []
    dfs(list(range(1, n + 1)), k, [], res)
    return res


def dfs(nums, k, path, res):
    if len(path) == k:
        res.append(path)
        return res
    for i in range(len(nums)):
        dfs(nums[i + 1 :], k, path + [nums[i]], res)


n = 4
k = 2
print(combine(n, k))

File: 3.math-stats/all-combinations/combine.rb
def combine(n, k)
  res = []
  nums = [*1..n]
  dfs(nums, k, [], res)
  return res
end

def dfs(nums, k, path, res)
  if path.length == k
    res.push(path)
    return res
  end
  for i in 0..(nums.length - 1)
    dfs(nums[i + 1..-1], k, path + [nums[i]], res)
  end
end

p(combine(4, 2))
